### ISSUE TITLE
validation should fail if instance types are unavailable in the given region

### ISSUE THREAD

**Issue by FirefoxMetzger** [hashicorp/terraform-provider-aws#31754]
### Description

I'm just starting out with Terraform (looking awesome so far 🚀 ) so this could very well be my naivete towards it that is the issue.

I was following the tutorial on getting started (https://developer.hashicorp.com/terraform/tutorials/aws-get-started/aws-build) which worked great, and afterwards wanted to modify it to use an AWS region closer to me (`eu-north-1`). I updated the region and the AMI accordingly:

```
provider "aws" {
  region  = "eu-north-1"
}

resource "aws_instance" "app_server" {
  ami           = "ami-0989fb15ce71ba39e"
  instance_type = "t2.micro"

  tags = {
    Name = "ExampleAppServerInstance"
  }
}
```

but the result kept failing with a (for me as a beginner) cryptic error message:

```
aws_instance.app_server: Creating...
╷
│ Error: creating EC2 Instance: Unsupported: The requested configuration is currently not supported. Please check the documentation for supported configurations.
│       status code: 400, request id: 96f6279e-6162-45a1-b5c3-026aae1bb2c2
│
│   with aws_instance.app_server,
│   on main.tf line 16, in resource "aws_instance" "app_server":
│   16: resource "aws_instance" "app_server" {
│
╵
```

`terraform validate` can't find an issue (`Success! The configuration is valid.`), and after spending about 1h searching the web I realized that **the problem is that `eu-north-1` doesn't offer `t2` instances. Instead, I have to use a `t3.micro`**.

---

Long story short, I think it would be really cool if `terraform validate` could warn/fail in this case and tell me that the chosen `instance_type` is unavailable for the chosen `region`.

Would this be a useful feature, or is this already available elsewhere?

### Would you like to implement a fix?

None

**Comment by justinretzolk** [comment]
Hey @FirefoxMetzger 👋 Thank you for taking the time to raise this, and I'm glad to hear you're enjoying Terraform so far! I'd like to leave this open as an enhancement request so that the maintainers can take a look when they can prioritize this to see if it's something that we may be able to build in. In the meantime, you could use the [`aws_ec2_instance_type_offerings` data source](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/ec2_instance_type_offerings) to make sure that the requested instance type is available in the given region/availability zone/etc.

**Comment by barnabyibis** [comment]
Interesting I should stumble across this while looking for a solution to validating an input variable to declare the AWS instance_type
`variable "default_instance" {
	type = string
	default = "t2.nano"
	validation {
		condition = contains(data.aws_ec2_instance_type_offerings.valid.instance_types,var.default_instance)
		error_message = "Supplied default instance is not a known valid type"
	}
}`
This throws the error 

> The condition for variable "default_instance" can only refer to the variable itself, using var.default_instance.
So I'm suggesting to either upvote the original concern, or propose implementing a solution to allow this workaround by pre-loading data blocks (and locals?) and allow them to help evaluate validation conditions.
